backend/middleware/audit/audit.go

```diff
@@ -86,23 +94,39 @@ func (m *mid) eventFromRequest(ctx context.Context, req interface{}, info *grpc.
 		username = claims.Subject
 	}
 
+	reqBody, err := meta.APIBody(req)
+	if err != nil {
+		return nil, err
+	}
+
 	return &auditv1.RequestEvent{
 		Username:    username,
 		ServiceName: svc,
 		MethodName:  method,
 		Type:        meta.GetAction(info.FullMethod),
 		Resources:   meta.ResourceNames(req.(proto.Message)),
-	}
+		RequestMetadata: &auditv1.RequestMetadata{
+			Body: reqBody,
+		},
+	}, nil
 }
 
-func (m *mid) eventFromResponse(resp interface{}, err error) *auditv1.RequestEvent {
+func (m *mid) eventFromResponse(resp interface{}, err error) (*auditv1.RequestEvent, error) {
 	s := status.Convert(err)
 	if s == nil {
 		s = status.New(codes.OK, "")
 	}
 
+	respBody, err := meta.APIBody(resp)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
IMO the biggest mistake we made the first time around was trying to record the body if err != nil. the primary return value should be considered unreliable in the case that an error occurred and discarded.

the rest of the checks are still potentially valuable to prevent a panic in the event that someone decides to `return nil, nil`, though as long as they never do it from middleware it wouldn't matter because `return nil, nil` from a handler would still be (*myProtoType)(nil) which the new any marshaller is fine with.

if err != nil let's ignore the resp altogether.

Reply:
in addition to the above and based on the scenario we tested yesterday (snippet below), the newest changes also do not call [`meta.ResourceNames`](https://github.com/lyft/clutch/blob/eec53c69d8c8511707fffae8ea3ff776a2c249ae/backend/middleware/audit/audit.go#L106) if err != nil. Currently, if resp is an untyped nil, `meta.ResourceNames` will panic with: `panic: interface conversion: interface is nil, not protoreflect.ProtoMessage`.
```
func (m *mid) UnaryInterceptor() grpc.UnaryServerInterceptor {
	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
		return nil, fmt.Errorf("problems")
	}
}
```

Reply:
with the latest changes (if UnaryInterceptor returned the snippet above)

<img src="https://user-images.githubusercontent.com/39421794/97344878-bd310c00-185f-11eb-997e-4ea1eceb4e39.png" alt="" width="300"/>